### ISSUE TITLE
Double softmax in MoE router load-balancing loss (mixtral, qwen2_moe, qwen3_vl_moe families)

### ISSUE THREAD

**Issue by ionut-anghelina** [huggingface/transformers#45120]
## Bug description

Several MoE routers apply `softmax` to raw logits inside their `forward()` method, then return the result as the first value (`router_logits`). This value is captured by `OutputRecorder` and passed to `load_balancing_loss_func`, which applies `softmax` **again** — computing the auxiliary loss on `softmax(softmax(logits))`.

This flattens the routing probability distribution toward uniform (`1/num_experts`), making `router_prob_per_expert` nearly constant regardless of actual routing decisions. The load-balancing loss effectively provides no useful gradient signal.

## Reproduction

```python
import torch
from transformers.models.qwen3_5_moe.configuration_qwen3_5_moe import Qwen3_5MoeTextConfig
from transformers.models.qwen3_5_moe.modeling_qwen3_5_moe import Qwen3_5MoeForCausalLM

config = Qwen3_5MoeTextConfig(
    vocab_size=1000, hidden_size=128, num_hidden_layers=4,
    num_attention_heads=4, num_key_value_heads=2, head_dim=32,
    moe_intermediate_size=64, shared_expert_intermediate_size=64,
    num_experts=8, num_experts_per_tok=2,
    linear_num_key_heads=4, linear_num_value_heads=4,
    linear_key_head_dim=32, linear_value_head_dim=32,
    linear_conv_kernel_dim=4, output_router_logits=True,
    router_aux_loss_coef=0.001, max_position_embeddings=128,
)

model = Qwen3_5MoeForCausalLM(config)
model.train()

input_ids = torch.randint(0, 1000, (1, 16))
outputs = model(input_ids, labels=input_ids.clone(), output_router_logits=True)

# router_logits are already probabilities (sum to 1.0 per row)
gate_probs = outputs.router_logits[0]
print("Row sums (should NOT be ~1.0 if these were raw logits):")
print(gate_probs.sum(dim=-1)[:5])
# Output: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], ...)
```

## Root cause

Taking `Qwen3_5MoeTopKRouter` as an example (`modeling_qwen3_5_moe.py`):

```python
def forward(self, hidden_states):
    router_logits = F.linear(hidden_states, self.weight)          # raw logits
    router_logits = torch.nn.functional.softmax(router_logits, dtype=torch.float, dim=-1)  # now probabilities!
    router_top_value, router_indices = torch.topk(router_logits, self.top_k, dim=-1)
    ...
    return router_logits, router_scores, router_indices  # returns probs as "logits"
```

Then in `load_balancing_loss_func`:
```python
routing_weights = torch.nn.functional.softmax(concatenated_gate_logits, dim=-1)  # softmax applied AGAIN
```

Note: the `SparseMoeBlock.forward()` discards the first return value (`_`), so only `OutputRecorder` → `load_balancing_loss_func` consumes it. No downstream routing logic is affected.

## Affected models

Three source routers (in modular files):
- `MixtralTopKRouter` in `mixtral/modular_mixtral.py`
- `Qwen2MoeTopKRouter` in `qwen2_moe/modular_qwen2_moe.py`
- `Qwen3VLMoeTextTopKRouter` in `qwen3_vl_moe/modular_qwen3_vl_moe.py`

Downstream models inheriting these routers: minimax, olmoe, flex_olmo, qwen3_moe, qwen3_next, qwen3_omni_moe, qwen3_5_moe

## Suggested fix

Use a separate variable for the softmaxed values, keeping `router_logits` as raw logits:

```python
router_logits = F.linear(hidden_states, self.weight)
router_probs = torch.nn.functional.softmax(router_logits, dtype=torch.float, dim=-1)
router_top_value, router_indices = torch.topk(router_probs, self.top_k, dim=-1)
...
return router_logits, router_scores, router_indices  # now returns actual logits
```

## Impact

This only affects **fine-tuning** with `output_router_logits=True`. It does not affect inference or pretrained model quality (those were trained with the original upstream codebases, not HuggingFace transformers).

**Comment by Whatsonyourmind** [comment]
Great catch. The math confirms this is a real bug with measurable impact.

When softmax is applied twice, the output distribution entropy increases toward `log(num_experts)`. For 8 experts with uniform-ish logits:

```
softmax([1,1,1,1,1,1,1,1]) = [0.125, ..., 0.125]  (uniform)
softmax(softmax([2,1,1,1,1,1,1,1])) ≈ [0.133, 0.124, ..., 0.124]
```

The double-softmax compresses the ratio between max and min routing probabilities. If the original logits produce probabilities `p_max / p_min = r`, then after double softmax:

```
effective_ratio ≈ exp(p_max - p_min) ≈ 1 + (p_max - p_min)  for small differences
```

Since softmax outputs are already in [0, 1] and close together, the second softmax nearly uniformizes them. This means `router_prob_per_expert` in the load-balancing loss becomes approximately `1/num_experts` regardless of actual routing, so the gradient of the auxiliary loss w.r.t. the router weights vanishes.

The fix should return raw logits from `forward()` and keep the softmax only in the routing path:

```python
def forward(self, hidden_states):
    router_logits = F.linear(hidden_states, self.weight)  # raw logits
    routing_probs = F.softmax(router_logits, dtype=torch.float, dim=-1)
    router_top_value, router_indices = torch.topk(routing_probs, self.top_k, dim=-1)
    ...
    return router_logits, router_scores, router_indices  # return RAW logits
```

**Comment by Rocketknight1** [comment]
Hey, this issue has gotten a little chaotic! It's been reported at #43542 I believe but we also have #45131. I'm going to try to push one of those fixes through today

**Comment by ionut-anghelina** [comment]
https://github.com/huggingface/transformers/pull/45111/changes
Opened a PR with the same fix over a week ago

**Comment by Rocketknight1** [comment]
@ionut-anghelina you did, and I'm sorry! Unfortunately, we're getting swamped by code agent PRs and we're rapidly closing a lot of them. Sometimes that means we do close legit fixes.

**Comment by ionut-anghelina** [comment]
So should i polish my PR a bit so it gets merged?

**Comment by Rocketknight1** [comment]
@ionut-anghelina yeah sure, it'll save me getting internal review - can you check to see the files that are different between your PR and mine, and if we need to copy those changes over, then ping me and I'll reopen + merge your one. Sorry for the chaos, we're just all very confused and overworked because of agent PRs 😅

**Comment by ionut-anghelina** [comment]
No problem. will check now

**Comment by ionut-anghelina** [comment]
https://github.com/huggingface/transformers/pull/45346/changes
@Rocketknight1

**Comment by vasqu** [comment]
Fixed by #45131
